pkg/bff/auth/cache/cache.go

```diff
@@ -0,0 +1,88 @@
+package cache
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/trisacrypto/directory/pkg/bff/config"
+)
+
+type ResourceFetcher interface {
+	Get(key string) (interface{}, error)
+}
+
+// TTLCache is a time-to-live cache that relies on a key value store for persistence,
+// automatically refreshing entries when they expire and uses a FIFO eviction policy
+// to manage the cache size.
+type TTLCache struct {
+	sync.RWMutex
+	fetcher ResourceFetcher
+	store   KeyValueStore
+	conf    config.CacheConfig
+	keys    []string
+}
+
+// NewUserCache creates a new user cache.
+func NewTTLCache(conf config.CacheConfig) *TTLCache {
+	return &TTLCache{
+		conf: conf,
+		keys: make([]string, 0),
+	}
+}
+
+func (c *TTLCache) SetFetcher(fetcher ResourceFetcher) {
+	c.fetcher = fetcher
+}
+
+func (c *TTLCache) SetStore(store KeyValueStore) {
+	c.store = store
+}
+
+// Get returns a value by key from the cache. If the key is not found or the value has
+// expired, we have to get a new value from the fetcher and handle evictions to abide
+// by the cache size limit.
+func (c *TTLCache) Get(key string) (value interface{}, err error) {
```

Comment:
This is a critical part of the code and it's where we would see race conditions/deadlocks so it's important that it's vetted closely.